File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    if(!digits.length) return [];
    const map = new Map([
        ['2','abc'],
        ['3','def'],
        ['4','ghi'],
        ['5','jkl'],
        ['6','mno'],
        ['7','pqrs'],
        ['8','tuv'],
        ['9','wxyz']
    ])
    const ans = [];
    let temp = "";
    const search = function(k){
        if(k==digits.length) ans.push(temp);
        else{
            const str = map.get(digits[k]);
            for(let choice of [...str]){
                temp += choice;
                search(k+1);
                temp = temp.slice(0,-1);
            }
        }
    }
    search(0);
    return ans;
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
private:
    vector<string> ans;
    string temp = "";
public:
    void search(string digits, int k){
        if(k == digits.size()){
            // goal
            ans.push_back(temp);
        }else{
            int num = digits[k] - '0';
            int end = (num==7 || num ==9) ? 4:3;
            for(int i = 0; i< end; i++){
                char choice = 'a' + (num-2)*3 + i;
                if(num == 8 || num == 9) choice += 1;
                temp.push_back(choice);
                search(digits,k+1);
                temp.pop_back();
            }
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.size() == 0) return ans;
        search(digits,0);
        return ans;
    }
};